File: src/ros_template_node/src/performance_monitor.cpp
#include "ros_template_node/performance_monitor.hpp"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace ros_template_node {

PerformanceMonitor::PerformanceMonitor(rclcpp::Node* node)
    : node_(node),
      start_time_(std::chrono::steady_clock::now()),
      last_metrics_time_(std::chrono::steady_clock::now())
{
  // Create metrics publisher
  metrics_publisher_ =
      node_->create_publisher<diagnostic_msgs::msg::DiagnosticArray>(
          "performance_metrics", 10);

  // Create metrics timer (10 second intervals)
  metrics_timer_ = node_->create_wall_timer(
      std::chrono::seconds(10), [this]() { this->publish_metrics(); });
}

void PerformanceMonitor::record_publish(const std::string& topic)
{
  std::lock_guard<std::mutex> lock(metrics_mutex_);
  total_messages_published_++;
  topic_publish_counts_[topic]++;
}

void PerformanceMonitor::record_callback_start()
{
  callback_start_time_ = std::chrono::steady_clock::now();
}

void PerformanceMonitor::record_callback_end()
{
  auto end_time = std::chrono::steady_clock::now();
  auto duration = std::chrono::duration_cast<std::chrono::microseconds>(
                      end_time - callback_start_time_)
                      .count() /
                  1000.0;  // Convert to milliseconds

  std::lock_guard<std::mutex> lock(metrics_mutex_);
  callback_times_.push_back(duration);

  // Keep only last 100 measurements
  if (callback_times_.size() > 100) {
    callback_times_.pop_front();
  }
}

void PerformanceMonitor::record_message_received(const std::string& topic)
{
  std::lock_guard<std::mutex> lock(metrics_mutex_);
  total_messages_received_++;
  topic_receive_counts_[topic]++;
}

double PerformanceMonitor::get_cpu_usage()
{
  static std::chrono::steady_clock::time_point last_time =
      std::chrono::steady_clock::now();
  static long long last_total_time = 0;
  static long long last_idle_time = 0;

  std::ifstream stat_file("/proc/stat");
  std::string line;
  if (!std::getline(stat_file, line)) {
    return 0.0;
  }

  std::istringstream iss(line);
  std::string cpu;
  long long user, nice, system, idle, iowait, irq, softirq, steal;

  iss >> cpu >> user >> nice >> system >> idle >> iowait >> irq >> softirq >>
      steal;

  long long total_time =
      user + nice + system + idle + iowait + irq + softirq + steal;
  long long current_idle = idle + iowait;

  auto current_time = std::chrono::steady_clock::now();
  auto time_diff = std::chrono::duration_cast<std::chrono::milliseconds>(
                       current_time - last_time)
                       .count();

  double cpu_percent = 0.0;
  if (time_diff > 500 &&
      last_total_time > 0) {  // Only calculate if >500ms passed
    long long total_diff = total_time - last_total_time;
    long long idle_diff = current_idle - last_idle_time;

    if (total_diff > 0) {
      cpu_percent = 100.0 * (total_diff - idle_diff) / total_diff;
    }

    last_time = current_time;
    last_total_time = total_time;
    last_idle_time = current_idle;
  }

  return std::max(0.0, std::min(100.0, cpu_percent));
}

size_t PerformanceMonitor::get_memory_usage_mb()
{
  std::ifstream status_file("/proc/self/status");
  std::string line;

  while (std::getline(status_file, line)) {
    if (line.find("VmRSS:") == 0) {
      std::istringstream iss(line);
      std::string label, value, unit;
      iss >> label >> value >> unit;
      return std::stoull(value) / 1024;  // Convert KB to MB
    }
  }
  return 0;
}

double PerformanceMonitor::get_uptime_seconds()
{
  auto now = std::chrono::steady_clock::now();
  return std::chrono::duration_cast<std::chrono::seconds>(now - start_time_)
      .count();
}

double PerformanceMonitor::calculate_average_callback_time()
{
  std::lock_guard<std::mutex> lock(metrics_mutex_);
  if (callback_times_.empty()) return 0.0;

  double sum = 0.0;
  for (double time : callback_times_) {
    sum += time;
  }
  return sum / callback_times_.size();
}

double PerformanceMonitor::calculate_max_callback_time()
{
  std::lock_guard<std::mutex> lock(metrics_mutex_);
  if (callback_times_.empty()) return 0.0;

  return *std::max_element(callback_times_.begin(), callback_times_.end());
}

double PerformanceMonitor::calculate_publish_rate()
{
  auto current_time = std::chrono::steady_clock::now();
  auto time_diff = std::chrono::duration_cast<std::chrono::seconds>(
                       current_time - last_metrics_time_)
                       .count();

  if (time_diff <= 0) return 0.0;

  size_t current_count = total_messages_published_.load();
  double rate =
      static_cast<double>(current_count - last_published_count_) / time_diff;

  last_metrics_time_ = current_time;
  last_published_count_ = current_count;

  return rate;
}

nlohmann::json PerformanceMonitor::get_metrics_json()
{
  std::lock_guard<std::mutex> lock(metrics_mutex_);

  nlohmann::json metrics;
  metrics["system"]["cpu_percent"] = get_cpu_usage();
  metrics["system"]["memory_mb"] = get_memory_usage_mb();
  metrics["system"]["uptime_seconds"] = get_uptime_seconds();

  metrics["ros"]["messages_published_total"] = total_messages_published_.load();
  metrics["ros"]["messages_received_total"] = total_messages_received_.load();
  metrics["ros"]["publish_rate_hz"] = calculate_publish_rate();
  metrics["ros"]["avg_callback_ms"] = calculate_average_callback_time();
  metrics["ros"]["max_callback_ms"] = calculate_max_callback_time();

  // Topic-specific metrics
  nlohmann::json topics;
  for (const auto& [topic, count] : topic_publish_counts_) {
    topics[topic]["published_count"] = count;
  }
  for (const auto& [topic, count] : topic_receive_counts_) {
    topics[topic]["received_count"] = count;
  }
  metrics["topics"] = topics;

  return metrics;
}

void PerformanceMonitor::publish_metrics()
{
  auto metrics_json = get_metrics_json();

  // Create diagnostic message
  auto diagnostic_array = diagnostic_msgs::msg::DiagnosticArray();
  diagnostic_array.header.stamp = node_->get_clock()->now();

  auto status = diagnostic_msgs::msg::DiagnosticStatus();
  status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
  status.name = std::string(node_->get_name()) + "_performance";
  status.message = "Performance metrics";
  status.hardware_id = "performance_monitor";

  // Add key metrics as key-value pairs
  auto add_kv = [&](const std::string& key, const std::string& value) {
    diagnostic_msgs::msg::KeyValue kv;
    kv.key = key;
    kv.value = value;
    status.values.push_back(kv);
  };

  add_kv("cpu_percent", std::to_string(static_cast<double>(
                            metrics_json["system"]["cpu_percent"])));
  add_kv(
      "memory_mb",
      std::to_string(static_cast<size_t>(metrics_json["system"]["memory_mb"])));
  add_kv("uptime_seconds", std::to_string(static_cast<double>(
                               metrics_json["system"]["uptime_seconds"])));
  add_kv("messages_published_total",
         std::to_string(static_cast<size_t>(
             metrics_json["ros"]["messages_published_total"])));
  add_kv("publish_rate_hz", std::to_string(static_cast<double>(
                                metrics_json["ros"]["publish_rate_hz"])));
  add_kv("avg_callback_ms", std::to_string(static_cast<double>(
                                metrics_json["ros"]["avg_callback_ms"])));

  diagnostic_array.status.push_back(status);
  metrics_publisher_->publish(diagnostic_array);
}

}  // namespace ros_template_node

File: src/py_template_node/py_template_node/performance_monitor.py
#!/usr/bin/env python3
"""Performance monitoring for ROS 2 Python nodes."""

import time
from collections import defaultdict, deque
from typing import Dict, Any, Optional
import threading

import psutil
import rclpy
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue
from rclpy.node import Node


class PerformanceMonitor:
    """Performance monitoring for ROS 2 Python nodes."""

    def __init__(self, node: Node):
        """Initialize the performance monitor.
        
        Args:
            node: The ROS 2 node to monitor
        """
        self.node = node
        self.start_time = time.time()
        self.last_metrics_time = time.time()
        
        # Message counters
        self.total_messages_published = 0
        self.total_messages_received = 0
        self.topic_publish_counts = defaultdict(int)
        self.topic_receive_counts = defaultdict(int)
        self.last_published_count = 0
        
        # Callback timing
        self.callback_times = deque(maxlen=100)
        self.callback_start_time: Optional[float] = None
        
        # Thread safety
        self.lock = threading.Lock()
        
        # System monitoring
        self.process = psutil.Process()
        
        # ROS integration
        self.metrics_publisher = self.node.create_publisher(
            DiagnosticArray, 'performance_metrics', 10
        )
        
        # Create metrics timer (10 second intervals)
        self.metrics_timer = self.node.create_timer(10.0, self.publish_metrics)

    def record_publish(self, topic: str) -> None:
        """Record a message publication.
        
        Args:
            topic: The topic name where message was published
        """
        with self.lock:
            self.total_messages_published += 1
            self.topic_publish_counts[topic] += 1

    def record_callback_start(self) -> None:
        """Record the start of a callback execution."""
        self.callback_start_time = time.time()

    def record_callback_end(self) -> None:
        """Record the end of a callback execution."""
        if self.callback_start_time is not None:
            duration_ms = (time.time() - self.callback_start_time) * 1000
            with self.lock:
                self.callback_times.append(duration_ms)
            self.callback_start_time = None

    def record_message_received(self, topic: str) -> None:
        """Record a message reception.
        
        Args:
            topic: The topic name where message was received
        """
        with self.lock:
            self.total_messages_received += 1
            self.topic_receive_counts[topic] += 1

    def get_system_metrics(self) -> Dict[str, float]:
        """Get current system metrics.
        
        Returns:
            Dictionary containing CPU, memory, and uptime metrics
        """
        try:
            cpu_percent = self.process.cpu_percent()
            memory_info = self.process.memory_info()
            memory_mb = memory_info.rss / 1024 / 1024  # Convert bytes to MB
            uptime_seconds = time.time() - self.start_time
            
            return {
                'cpu_percent': cpu_percent,
                'memory_mb': memory_mb,
                'uptime_seconds': uptime_seconds
            }
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            return {
                'cpu_percent': 0.0,
                'memory_mb': 0.0,
                'uptime_seconds': time.time() - self.start_time
            }

    def get_ros_metrics(self) -> Dict[str, float]:
        """Get ROS-specific performance metrics.
        
        Returns:
            Dictionary containing ROS message and timing metrics
        """
        with self.lock:
            # Calculate publish rate
            current_time = time.time()
            time_diff = current_time - self.last_metrics_time
            
            if time_diff > 0:
                publish_rate = (self.total_messages_published - self.last_published_count) / time_diff
                self.last_metrics_time = current_time
                self.last_published_count = self.total_messages_published
            else:
                publish_rate = 0.0
            
            # Calculate callback timing statistics
            avg_callback_ms = 0.0
            max_callback_ms = 0.0
            if self.callback_times:
                avg_callback_ms = sum(self.callback_times) / len(self.callback_times)
                max_callback_ms = max(self.callback_times)
            
            return {
                'messages_published_total': self.total_messages_published,
                'messages_received_total': self.total_messages_received,
                'publish_rate_hz': publish_rate,
                'avg_callback_ms': avg_callback_ms,
                'max_callback_ms': max_callback_ms
            }

    def get_topic_metrics(self) -> Dict[str, Dict[str, int]]:
        """Get per-topic metrics.
        
        Returns:
            Dictionary containing per-topic message counts
        """
        with self.lock:
            topics = {}
            
            # Combine publish and receive counts
            all_topics = set(self.topic_publish_counts.keys()) | set(self.topic_receive_counts.keys())
            
            for topic in all_topics:
                topics[topic] = {
                    'published_count': self.topic_publish_counts.get(topic, 0),
                    'received_count': self.topic_receive_counts.get(topic, 0)
                }
            
            return topics

    def get_metrics_dict(self) -> Dict[str, Any]:
        """Get all metrics as a structured dictionary.
        
        Returns:
            Complete metrics dictionary for JSON serialization
        """
        return {
            'system': self.get_system_metrics(),
            'ros': self.get_ros_metrics(),
            'topics': self.get_topic_metrics()
        }

    def publish_metrics(self) -> None:
        """Publish performance metrics as ROS diagnostic messages."""
        metrics = self.get_metrics_dict()
        
        # Create diagnostic array message
        diagnostic_array = DiagnosticArray()
        diagnostic_array.header.stamp = self.node.get_clock().now().to_msg()
        
        # Create performance status
        status = DiagnosticStatus()
        status.level = DiagnosticStatus.OK
        status.name = f"{self.node.get_name()}_performance"
        status.message = "Performance metrics"
        status.hardware_id = "performance_monitor"
        
        # Add key metrics as key-value pairs
        def add_kv(key: str, value: Any) -> None:
            kv = KeyValue()
            kv.key = key
            kv.value = str(value)
            status.values.append(kv)
        
        # System metrics
        system_metrics = metrics['system']
        add_kv('cpu_percent', f"{system_metrics['cpu_percent']:.2f}")
        add_kv('memory_mb', f"{system_metrics['memory_mb']:.2f}")
        add_kv('uptime_seconds', f"{system_metrics['uptime_seconds']:.1f}")
        
        # ROS metrics
        ros_metrics = metrics['ros']
        add_kv('messages_published_total', ros_metrics['messages_published_total'])
        add_kv('messages_received_total', ros_metrics['messages_received_total'])
        add_kv('publish_rate_hz', f"{ros_metrics['publish_rate_hz']:.2f}")
        add_kv('avg_callback_ms', f"{ros_metrics['avg_callback_ms']:.2f}")
        add_kv('max_callback_ms', f"{ros_metrics['max_callback_ms']:.2f}")
        
        diagnostic_array.status.append(status)
        self.metrics_publisher.publish(diagnostic_array)
        
        # Log metrics summary periodically
        self.node.get_logger().info(
            f"Performance: CPU {system_metrics['cpu_percent']:.1f}%, "
            f"Memory {system_metrics['memory_mb']:.1f}MB, "
            f"Rate {ros_metrics['publish_rate_hz']:.1f}Hz, "
            f"Avg callback {ros_metrics['avg_callback_ms']:.1f}ms"
        )